### PR TITLE
Adds migration resource

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/2652
Towards https://github.com/giantswarm/giantswarm/issues/2947

### CODE REVIEW

File: service/controller/v11/resource/migration/resource.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	if credentialSecret.Namespace == "" || credentialSecret.Name == "" {
+		r.logger.Log("level", "debug", "message", "config is missing default credential, migrating")
+
+		customObject.Spec.AWS.CredentialSecret.Namespace = CredentialSecretDefaultNamespace
+		customObject.Spec.AWS.CredentialSecret.Name = CredentialSecretDefaultName
+
+		_, err := r.g8sClient.ProviderV1alpha1().AWSConfigs(AWSConfigNamespace).Update(&customObject)
+		if err != nil {
+			return microerror.Mask(err)
+		}
```
Comment (JosephSalisbury) [inline service/controller/v11/resource/migration/resource.go]: TODO: Cancel reconciliation here.

File: service/controller/v11/cluster_resource_set.go [hunk 2]
```diff
@@ -63,6 +66,9 @@ func NewClusterResourceSet(config ClusterResourceSetConfig) (*controller.Resourc
 	if config.CertsSearcher == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.CertsSearcher must not be empty")
 	}
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.G8sClient must not be empty")
+	}
```
Comment (rossf7) [inline service/controller/v11/cluster_resource_set.go]: Can we use the `%T.G8sClient` foo here? I find it easier to read in errors.

File: service/controller/v11/resource/migration/resource.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// If credential secret is set we have nothing to migrate.
+	if customObject.Spec.AWS.CredentialSecret.Name != "" {
+		return nil
+	}
```
Comment (rossf7) [inline service/controller/v11/resource/migration/resource.go]: Will this be used outside of the migration resource? If so then a key function would be cool.

Comment (kopiczko) [reply service/controller/v11/resource/migration/resource.go]: I have mixed feeling about that. We don't have key function for setting things and this value is set a few lines later.

This is a migration resource so I don't see anything wrong with it knowing the structure.

File: service/controller/v11/cluster_resource_set.go (hunk 2 above)
Comment (kopiczko) [reply service/controller/v11/cluster_resource_set.go]: Let's do that in a separate PR. This is just a copy/paste from the previous line.

File: service/controller/v11/resource/migration/resource.go (hunk 3 above)
Comment (rossf7) [reply service/controller/v11/resource/migration/resource.go]: Yes, good point about setting the value. If this is just used within the migration resource I think its fine without it.

If we need to read the value in other resources a key function can be added later.

File: service/controller/v11/resource/migration/resource.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Package migration provides an operatorkit resource that migrates awsconfig CRs
+// to reference the default credential secret if they do not already.
+// It can be safely removed once all awsconfig CRs reference a credential secret.
```
Comment (xh3b4sd) [inline service/controller/v11/resource/migration/resource.go]: I do not understand the migration thingy. Why do we need it? I would expect customers using the BYOC magic in production right now. I am not aware of this. So what do we need to migrate at all? Also, what does `to reference the default credential secret` actually mean?

File: service/controller/v11/resource/migration/resource.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+package migration
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/aws-operator/service/controller/v11/key"
```
Comment (xh3b4sd) [inline service/controller/v11/resource/migration/resource.go]: Separate block would be cool.

File: service/controller/v11/resource/migration/resource.go (hunk 4 above)
Comment (kopiczko) [reply service/controller/v11/resource/migration/resource.go]: We need it so this change https://github.com/giantswarm/aws-operator/pull/916 can actually work. If you check our installations we have `giantswarm/credential-default` secret everywhere. It's prepared to allow access to current guest AWS accounts. In the future (as soon as v10 and predecessors are removed) we'll remove guest AWS credential flags.

File: service/controller/v11/resource/migration/resource.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+)
+
+const (
+	Name = "migrationv11"
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/v11/resource/migration/resource.go]: No, we don't. They are private now.

File: service/controller/v11/resource/migration/resource.go (hunk 5 above)
Comment (kopiczko) [reply service/controller/v11/resource/migration/resource.go]: Done.

**Comment by kopiczko** [comment]
e2e were green. We are ready to merge mr @JosephSalisbury.

**Pull request merged**